### PR TITLE
  TxSim: Transaction Simulator and Risk Detector for Polkadot

### PR DESCRIPTION
**Pull request by EmmanuelEklipse**
#### Project Abstract

> TxShield is a transaction simulator and risk detection tool designed to prevent failed transactions, gas waste, and honeypot scams by previewing potential risks before execution. Built for Polkadot using Next.js for the frontend, leveraging Node.js and the Polkadot.js API for backend transaction simulation and blockchain interactions. It uses custom WASM bytecode analysis for detecting risky contracts and integrates with wallets like Talisman and SubWallet to provide a seamless user experience.   Targeting Polkadot’s rapidly growing ecosystem of 2+ million wallet users and aims to expand support across multiple parachains for broader coverage and safer transactions. 
>

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (Polkadot AssetHub (USDC & DOT) address in the application and bank details via email, if applicable).
- [x] I understand that an agreed upon percentage of each milestone will be paid in vested DOT, to the Polkadot address listed in the application.
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Comment by EmmanuelEklipse** [comment]
I have read and hereby sign the Contributor License Agreement.

**Review by keeganquigley — CHANGES_REQUESTED** [review]
Hi @EmmanuelEklipse the PR correctly reflects a level 3 grant, but the application notes a level 2 grant. Can you update this on the application itself?

**Comment by EmmanuelEklipse** [comment]
@keeganquigley Thanks for pointing it out. I have corrected it

**Comment by keeganquigley** [comment]
Thanks @EmmanuelEklipse we usually only allow one grant application to be open at a time per grantee. Would you prefer to keep this one open and close https://github.com/w3f/Grants-Program/pull/2563 since it hasn't garnered enough approvals?

**Comment by EmmanuelEklipse** [comment]
@keeganquigley Yes, please.

**Comment by keeganquigley** [comment]
Thanks @EmmanuelEklipse I will mark as ready for review and ping the rest of the committee for comment. In the meantime, one more question:

Would you be willing to support upcoming PVM (Asset Hub) as well as EVM and WASM? Please see this [forum post](https://forum.polkadot.network/t/solidity-smart-contracts-striking-a-balance-between-innovating-and-traction/13661) regarding the latest news there.

**Comment by EmmanuelEklipse** [comment]
@keeganquigley Much appreciated.

Yes, we’re absolutely willing to support the upcoming PVM developments on Asset Hub — including both the WASM-based PolkaVM and the potential EVM-compatible stack. TxSim will be built with a modular architecture, and we’ve already accounted for multi-VM support (EVM via Moonbeam, WASM via Astar, and PVM via Asset Hub). This makes it straightforward for us to extend compatibility to Polkadot Hub's evolving smart contract layers without requiring additional funding, it can be delivered as a natural extension within our current technical framework.

We’re actively tracking the roadmap around PolkaVM, Solidity support, and EVM compatibility on the Hub. TxSim’s simulation and detection logic is designed to adapt to such changes, and we’re committed to ensuring it remains aligned with the most widely used execution environments in the Polkadot ecosystem.

**Comment by EmmanuelEklipse** [comment]
@keeganquigley Thanks for the clarification. Just updated the proposal to reflect the planned phased support for the RISC-V–based PVM, including the challenges and our modular approach. Please let me know if you have further feedback.

**Comment by EmmanuelEklipse** [comment]
Hi @keeganquigley any update on this?

**Comment by keeganquigley** [comment]
Hi @EmmanuelEklipse it seems like so far it hasn't been able to garner enough approvals. in my opinion I think it's a bit expensive for mainly a tx simulator. It might be hard to get 5 approvals for this. Would you perhaps be willing to lower it to a $30k level 2 PoC, even if its just for one milestone? Maybe you could hold off on the risk detection module until a future M2. Then you would only need 3 approvals and would give us a chance to see the simulator in action before spending more money on it.

**Comment by EmmanuelEklipse** [comment]
@keeganquigley Thanks for the feedback! I've updated the budget to $28K and revised the milestones accordingly. Let me know what you think or if you'd like to see any other adjustments.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks a lot for the application. It seems like you only focus on EVM as part of this grant. How are you planning to compete with established teams like https://www.hypernative.io/ that at this point also have a brand name and are used by various projects.

**Comment by EmmanuelEklipse** [comment]
Thanks for the question @N0c2! While Milestone 1 focuses on Moonbeam’s EVM compatibility for speed of delivery, TxSim is designed as a **multi-VM simulation framework**, meaning from the start it will support **WASM-based contracts on Astar** and runtime calls on **Asset Hub** (PVM) as the tooling matures.

This is one of our main differentiators from Hypernative, which is EVM-focused. Our competitive advantages are:

1. **Polkadot-native focus:** Simulation for Substrate-based WASM contracts and extrinsics, not just EVM.  
2. **Fee estimation for Polkadot’s weight-based model:**  This is a unique need for Substrate chains that EVM-first tools typically don’t address.  
3. **Wallet-first integration:**  Targeted integration with Talisman and SubWallet so the simulation and fee data are available *before signing*, directly in the wallet.  
4. **Public-good, open-source approach:** We’re making TxSim fully MIT-licensed, so any wallet, dApp, or parachain can integrate without vendor lock-in.

Our aim is not to replicate Hypernative’s EVM-only solution, but to **fill the gap for Polkadot’s diverse execution environments** and give users safer transaction previews across Moonbeam, Astar, and Asset Hub.

**Comment by EmmanuelEklipse** [comment]
@N0c2 Thanks for the suggestion. We've updated the milestones to make the WASM and pallet compatibility explicit. TxSim is being built as a multi-VM simulation framework, so Milestone 1 will support Moonbeam (EVM) and Astar (WASM) from the start, with the architecture ready for Substrate pallet/runtime extrinsic simulation (e.g., Asset Hub).

**Review by quigl1kd — DISMISSED** [review]
thx @EmmanuelEklipse for the changes, I'm now happy to approve it. While waiting for other reviews feel free to start the KYB verification process for the Build Union entity. Thanks!

**Review by keeganquigley — APPROVED** [review]
whoops sorry wrong account

**Comment by EmmanuelEklipse** [comment]
Thanks @quigl1kd @keeganquigley  for reviewing and approving the updates! 🙏 I’ll get started on the KYB verification process while we wait for the final approval.

File: applications/TxSim.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
+# TxSim: Transaction Simulator and Risk Detector for Polkadot
+
+**Team Name:** Build Union  
+**Payment Details:**
+
+- **DOT:** 1XYfhcsZFhcszyy6jiHYvkBqPxUXq6YADRUaUkcZ8vopqNTiqjBv
+- **USDC:** 0x53785b7155d016ba9c64fc2117d0eac0b21c9869
```
Comment (PieWol) [inline applications/TxSim.md:7]: This also needs to be a Polkadot AssetHub Address.

File: applications/TxSim.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+## Team
+
+**Build Union** is a collaborative collective of talented developers, designers, strategists, and innovators dedicated to shaping the future of the decentralized web. As an aggregate team of Web3 builders, Build Union brings together diverse expertise to create cutting-edge blockchain applications, DAO services, and decentralized solutions, focused on fostering innovation and accelerating Web3 adoption.
+
+### Members
+
+**Emmanuel — Project Lead**  
+Role: Project manager and blockchain architect with deep experience in DAO tooling and transparency infrastructure. Combines hands-on engineering with leadership in dApp development and open governance tooling.  
+LinkedIn: [linkedin.com/in/emmanuel-eclipse](https://linkedin.com/in/emmanuel-eclipse)
```
Comment (PieWol) [inline applications/TxSim.md:96]: this link doesn't seem to exist. Can you update it?

File: applications/TxSim.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
+# TxSim: Transaction Simulator and Risk Detector for Polkadot
+
+**Team Name:** Build Union  
+**Payment Details:**
+
+- **DOT:** 1XYfhcsZFhcszyy6jiHYvkBqPxUXq6YADRUaUkcZ8vopqNTiqjBv
+- **USDC:** 0x53785b7155d016ba9c64fc2117d0eac0b21c9869  
+  **Grant Level:** 3
```
Comment (PieWol) [inline applications/TxSim.md:8]:
```suggestion
  **Grant Level:** 2
```

**Review by PieWol — CHANGES_REQUESTED** [review]
Left some comments. Thanks for the application.

**Comment by EmmanuelEklipse** [comment]
@PieWol Done. Please check. I’m in the process of verifying the KYB and should have it all sorted by next tomorrow.

**Review by keeganquigley — APPROVED** [review]
re-approving

**Review by Noc2 — APPROVED** [review]

File: applications/TxSim.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+| 0a.    | License                | MIT                                                                                                                                            |
+| 0b.    | Documentation          | We will provide both inline documentation and a user guide explaining how to run the simulation engine, estimate fees, and interpret outputs. |
+| 0c.    | Testing and Guide      | Core functions will be covered by unit tests (≥80% coverage). We will include a guide to run and validate them.                               |
+| 0d.    | Docker                 | A Dockerfile will be provided for quick testing and deployment of the simulation engine.                                                      |
+| 1.     | Simulation Engine      | Initial transaction simulation engine targeting Moonbeam (EVM) and Astar (WASM) networks, with an architecture designed for future compatibility with Substrate pallets and runtime calls (e.g., Asset Hub PVM).                                                         |
```
Comment (PieWol) [inline applications/TxSim.md:159]: It might be more appropriate to not specify Astar in the deliverable, since they just use ``pallet-contracts`` which simply a pallet and well defines what we are talking about. Let me know what you think.

File: applications/TxSim.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+| 0a.    | License              | MIT                                                                                                                                        |
+| 0b.    | Documentation        | Updated documentation to reflect new features and enhancements.                                                                            |
+| 0c.    | Testing and Guide    | ≥90% test coverage with an expanded test suite to handle edge cases.                                                                      |
+| 0d.    | Docker               | Updated Dockerfile for the enhanced version of the simulation engine.                                                                      |
+| 1.     | Fee Optimization     | Advanced logic to recommend optimal transaction parameters (e.g., gas limits, max fees) based on simulation data.                         |
+| 2.     | Performance Tuning   | Enhancements in simulation speed and accuracy for an improved development experience.                                                                  |
```
Comment (PieWol) [inline applications/TxSim.md:179]: What are these enhancements? Sounds very unspecific.

File: applications/TxSim.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+|--------|----------------------|---------------------------------------------------------------------------------------------------------------------------------------------|
+| 0a.    | License              | MIT                                                                                                                                        |
+| 0b.    | Documentation        | Updated documentation to reflect new features and enhancements.                                                                            |
+| 0c.    | Testing and Guide    | ≥90% test coverage with an expanded test suite to handle edge cases.                                                                      |
+| 0d.    | Docker               | Updated Dockerfile for the enhanced version of the simulation engine.                                                                      |
+| 1.     | Fee Optimization     | Advanced logic to recommend optimal transaction parameters (e.g., gas limits, max fees) based on simulation data.                         |
```
Comment (PieWol) [inline applications/TxSim.md:178]: How can developers test their contracts with your tool, without already having them deployed on-chain somewhere?

**Review by PieWol — CHANGES_REQUESTED** [review]
Left a few more comments. Looking forward to your feedback.

File: applications/TxSim.md (hunk 4 above)
Comment (EmmanuelEklipse) [reply applications/TxSim.md:159]: Thanks @PieWol Astar was just used as an example. This depends on `pallet-contracts`, so the scope naturally extends to any chain supporting ink! contracts.

File: applications/TxSim.md (hunk 5 above)
Comment (EmmanuelEklipse) [reply applications/TxSim.md:179]: Sure, I can clarify the “Performance Tuning” deliverable. It refers to

Reducing average simulation runtime, aiming for ≥30% compared to the M1 baseline
Improving fee estimation accuracy to within ±5% deviation on benchmark test cases
With M2, our main goal is to add advanced functionality and improve usability and integration.

**Comment by EmmanuelEklipse** [comment]
KYB completed 
WhatsApp Image 2025-09-01 at 23 29 26_188fec2b

**Comment by EmmanuelEklipse** [comment]
@PieWol, bringing this to your attention. Please let me know if you need any further clarification.

**Comment by EmmanuelEklipse** [comment]
Hi @semuelle  @laboon  @takahser , KYB is done. We’ve got 2 approvals + 1 support, just one more to go. This tool lets developers simulate and validate (EVM + ink!) transactions before deployment, and end users will also benefit by seeing accurate fee estimates and risk checks before signing. It helps protect users from scams by catching issues early. Other networks have or are building similar tools, but Polkadot doesn't. Ours adds unique wallet integration that gives the project an edge; we’d love to bring this to life. Thanks for your support.

**Comment by keeganquigley** [comment]
Thanks @EmmanuelEklipse can confirm KYB is approved, I just pinged the team again.

**Comment by EmmanuelEklipse** [comment]
Thank you.

**Review by semuelle — CHANGES_REQUESTED** [review]
Just some minor remarks

**Review by Noc2 — APPROVED** [review]

**Review by keeganquigley — APPROVED** [review]
re-approving

**Comment by keeganquigley** [comment]
Congrats @EmmanuelEklipse the grant is now active!

**Comment by EmmanuelEklipse** [comment]
Thank you to everyone who has been a part of this 🎉. We’re really excited to see the grant officially active and to bring this project to fruition. It’s something the team fully believes in, and we’re looking forward to seeing it actively used by developers and the community.

**Comment by keeganquigley** [comment]
Hi @EmmanuelEklipse how is milestone 1 coming along? Can you please provide an update?

**Comment by EmmanuelEklipse** [comment]
Hi @keeganquigley , thanks for the nudge. We're behind schedule, but it is very much active. We are currently finalizing the core component simulation and expect to have the delivery posted within the next 2 weeks, with the remaining milestones to follow within a month. I will provide bi-weekly updates here until it's completed.

**Comment by EmmanuelEklipse** [comment]
Hi W3F Team,
I’m writing to provide an update on our progress for Milestone 1. While the core logic for the EVM/WASM simulation is largely implemented, we require an additional 1–2 weeks to further optimize and harden the codebase. Our goal is to ensure the system is lean and follows best practices for maintainability within the Polkadot ecosystem. We want Milestone 1 to include a robust, automated test suite with high branch coverage and a well-structured core engine. We believe this focus on technical robustness is essential for the project’s success and long-term value to the community.
We appreciate your flexibility as we complete these final refinements.
